engine/src/main/java/io/camunda/zeebe/engine/processing/common/ExpressionProcessor.java

```diff
@@ -186,23 +187,52 @@ public Either<Failure, Interval> evaluateIntervalExpression(
    */
   public Either<Failure, ZonedDateTime> evaluateDateTimeExpression(
       final Expression expression, final Long scopeKey) {
+    return evaluateDateTimeExpression(expression, scopeKey, false)
+        .flatMap(optionalDateTime -> Either.right(optionalDateTime.get()));
+  }
+
+  /**
+   * Evaluates the given expression and returns the result as <code>Optional<ZonedDateTime></code>.
+   * By using the <code>isNullable</code> flag, it is also possible to control the behavior if a
+   * given expression evaluates to an empty String or null. If the expression evaluation fails, the
+   * result is not a ZonedDateTime, or the result is not an empty String, then a failure is
+   * returned.
+   *
+   * @param expression the expression to evaluate
+   * @param scopeKey the scope to load the variables from (a negative key is intended to imply an
+   *     empty variable context)
+   * @param isNullable set to <code>true</code> to ignore empty strings instead of returning a
+   *     failure
+   * @return either the evaluation result as ZonedDateTime or a failure
+   * @throws EvaluationException if expression evaluation failed
+   */
+  public Either<Failure, Optional<ZonedDateTime>> evaluateDateTimeExpression(
+      final Expression expression, final Long scopeKey, final boolean isNullable) {
     final var result = evaluateExpression(expression, scopeKey);
     if (result.isFailure()) {
       return Either.left(
           new Failure(result.getFailureMessage(), ErrorType.EXTRACT_VALUE_ERROR, scopeKey));
     }
     if (result.getType() == ResultType.DATE_TIME) {
-      return Either.right(result.getDateTime());
+      return Either.right(Optional.of(result.getDateTime()));
     }
     if (result.getType() == ResultType.STRING) {
+
```

Comment:
Hey @berkaycanbc, I'm not sure I understood the part about the nullable guard clause.

What about moving the "nullable" logic into its separate `if` block without changing the `ResultType.STRING` if block? It would look like the following:

```java
if (isNullable && isResultNullOrEmptyString(result)) {
      return Either.right(Optional.empty());
}
```

Where the `isResultNullOrEmptyString` method would check if the `ResultType` is `NULL` or `STRING` with an empty `String`.

Reply:
I think, the last change definitely improved the readability. At the same time, I think, we can further improve readability by creating a private method and calling it inside as such:

```java
private Either<Failure, Optional<ZonedDateTime>> evaluateDateTimeString(
      final Expression expression, final Long scopeKey, final boolean isNullable) {
    final var result = evaluateExpression(expression, scopeKey);

    if (isNullable && (result.getString() == null || result.getString().isEmpty())) {
      return Either.right(Optional.empty());
    }

    final var resultString = result.getString();
    try {
      return Either.right(Optional.of(ZonedDateTime.parse(resultString)));
    } catch (final DateTimeParseException e) {
      return Either.left(
          new Failure(
              String.format(
                  "Invalid date-time format '%s' for expression '%s'",
                  resultString, expression.getExpression()),
              ErrorType.EXTRACT_VALUE_ERROR,
              scopeKey));
    }
  }
```

```java
    if (result.getType() == ResultType.STRING) {
      return evaluateDateTimeString(expression, scopeKey, isNullable);
    }
```

And remove the if check: `if (isNullable && isResultNullOrEmptyString(result))`

What do you think? @koevskinikola

Reply:
Hey @berkaycanbc, I've extracted the "string" section of the code in its own private method. There is a check for an empty String there.

There still is a guard condition in case we get a `ResultType.NULL` evaluation. As we discussed, this is a slight code duplication, but the readability of the code is improved, so we can accept it.

Reply:
Looks exactly what we intended. Thanks!